File: src/main/java/org/voegtle/weatherwidget/location/LocationView.java
package org.voegtle.weatherwidget.location;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.GridLayout;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;
import org.voegtle.weatherwidget.R;
import org.voegtle.weatherwidget.data.Statistics;
import org.voegtle.weatherwidget.data.StatisticsSet;
import org.voegtle.weatherwidget.data.WeatherData;
import org.voegtle.weatherwidget.util.DataFormatter;

public class LocationView extends LinearLayout {
  private final Context context;
  private TextView captionView;
  private ImageButton resizeButton;
  private ImageButton diagramButton;
  private ImageButton forecastButton;
  private GridLayout moreData;
  private TextView kwhCaptionView;

  private boolean expanded;
  private Drawable imageExpand;
  private Drawable imageCollapse;
  private OnClickListener externalClickListener;
  private OnClickListener diagramListener;
  private OnClickListener forecastListener;

  private DataFormatter formatter;


  public LocationView(Context context, AttributeSet attrs) {
    super(context, attrs);
    this.context = context;
    formatter = new DataFormatter();

    LayoutInflater li = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
    li.inflate(R.layout.view_location, this, true);

    captionView = (TextView) findViewById(R.id.caption);
    moreData = (GridLayout) findViewById(R.id.more_data);
    kwhCaptionView = (TextView) findViewById(R.id.caption_kwh);

    imageCollapse = context.getResources().getDrawable(R.drawable.ic_action_collapse);
    imageExpand = context.getResources().getDrawable(R.drawable.ic_action_expand);

    TypedArray attributes = context.getTheme().obtainStyledAttributes(
        attrs, R.styleable.LocationView, 0, 0);

    try {
      initializeTextViews(attributes);
      initializeButtons(attributes);
    } finally {
      attributes.recycle();
    }
  }

  private void initializeButtons(TypedArray attributes) {
    resizeButton = (ImageButton) findViewById(R.id.resize_button);
    resizeButton.setOnClickListener(new View.OnClickListener() {

      @Override
      public void onClick(View arg0) {
        setExpanded(!expanded);
        if (externalClickListener != null) {
          externalClickListener.onClick(arg0);
        }
      }
    });

    boolean diagrams = attributes.getBoolean(R.styleable.LocationView_diagrams, false);
    diagramButton = (ImageButton) findViewById(R.id.diagram_button);
    diagramButton.setVisibility(diagrams ? View.VISIBLE : View.GONE);
    diagramButton.setOnClickListener(new OnClickListener() {
      @Override
      public void onClick(View view) {
        if (diagramListener != null) {
          diagramListener.onClick(LocationView.this);
        }
      }
    });

    boolean forecast = attributes.getBoolean(R.styleable.LocationView_forecast, true);
    forecastButton = (ImageButton) findViewById(R.id.forecast_button);
    forecastButton.setVisibility(forecast ? View.VISIBLE : View.GONE);
    forecastButton.setOnClickListener(new OnClickListener() {
      @Override
      public void onClick(View view) {
        if (forecastListener != null)
          forecastListener.onClick(LocationView.this);
      }
    });

  }

  public void setOnClickListener(OnClickListener listener) {
    externalClickListener = listener;
  }

  public void setDiagramsOnClickListener(OnClickListener listener) {
    diagramListener = listener;
  }

  public void setForecastOnClickListener(OnClickListener listener) {
    forecastListener = listener;
  }

  private void initializeTextViews(TypedArray attributes) {
    String caption = attributes.getString(R.styleable.LocationView_caption);
    captionView.setText(caption);

  }

  public void setTextSize(int textSize) {
    captionView.setTextSize(textSize + 1);
    updateTextSize((GridLayout) findViewById(R.id.grid_current), textSize);
    updateTextSize((GridLayout) findViewById(R.id.more_data), textSize);
  }

  private void updateTextSize(GridLayout container, int textSize) {
    for (int i = 0; i < container.getChildCount(); i++) {
      TextView view = (TextView) container.getChildAt(i);
      view.setTextSize(textSize);
    }
  }


  public void setCaption(String caption) {
    captionView.setText(caption);
  }

  public void setData(WeatherData data) {
    TextView temperature = (TextView) findViewById(R.id.temperature);
    temperature.setText(formatter.formatTemperatureForActivity(data));

    TextView humidity = (TextView) findViewById(R.id.humidity);
    humidity.setText(formatter.formatPercent(data.getHumidity()));

    setRainData(data.getRain(), R.id.label_rain_last_hour, R.id.rain_last_hour);
    setRainData(data.getRainToday(), R.id.label_rain_today, R.id.rain_today);
    setSolarData(data.getWatt(), R.id.label_solar_output, R.id.solar_output);
  }

  private void setSolarData(Float watt, int labelId, int dataId) {
    TextView solarLabel = (TextView) findViewById(labelId);
    TextView solar = (TextView) findViewById(dataId);
    if (watt != null && watt != 0.0) {
      solarLabel.setVisibility(View.VISIBLE);
      solar.setVisibility(View.VISIBLE);
      solar.setText(formatter.formatOutput(watt));
    } else {
      solarLabel.setVisibility(View.GONE);
      solar.setVisibility(View.GONE);
    }
  }

  private void setRainData(Float value, int labelId, int dataId) {
    TextView rainLabel = (TextView) findViewById(labelId);
    TextView rain = (TextView) findViewById(dataId);
    if (value != null) {
      rainLabel.setVisibility(View.VISIBLE);
      rain.setVisibility(View.VISIBLE);
      rain.setText(formatter.formatRain(value));
    } else {
      rainLabel.setVisibility(View.GONE);
      rain.setVisibility(View.GONE);

    }
  }

  public void setMoreData(Statistics statistics) {
    if (statistics == null) {
      statistics = new Statistics();
    }

    StatisticsSet today = statistics.get(Statistics.TimeRange.today);
    updateStatistics(today, R.id.today_rain, R.id.today_min_temperature, R.id.today_max_temperature, R.id.today_kwh);

    StatisticsSet yesterday = statistics.get(Statistics.TimeRange.yesterday);
    updateStatistics(yesterday, R.id.yesterday_rain, R.id.yesterday_min_temperature, R.id.yesterday_max_temperature, R.id.yesterday_kwh);

    StatisticsSet week = statistics.get(Statistics.TimeRange.last7days);
    updateStatistics(week, R.id.week_rain, R.id.week_min_temperature, R.id.week_max_temperature, R.id.week_kwh);

    StatisticsSet month = statistics.get(Statistics.TimeRange.last30days);
    updateStatistics(month, R.id.month_rain, R.id.month_min_temperature, R.id.month_max_temperature, R.id.month_kwh);
  }

  private void updateStatistics(StatisticsSet stats, int rainId, int minTemperatureId, int maxTemperatureId, int kwhId) {
    TextView rainView = (TextView) findViewById(rainId);
    TextView minView = (TextView) findViewById(minTemperatureId);
    TextView maxView = (TextView) findViewById(maxTemperatureId);
    TextView kwhView = (TextView) findViewById(kwhId);
    rainView.setText("");
    minView.setText("");
    maxView.setText("");
    kwhView.setText("");
    if (stats != null) {
      rainView.setText(formatter.formatRain(stats.getRain()));
      minView.setText(formatter.formatTemperature(stats.getMinTemperature()));
      maxView.setText(formatter.formatTemperature(stats.getMaxTemperature()));
      if (stats.getKwh() != null) {
        kwhView.setText(formatter.formatKwh(stats.getKwh()));
        kwhCaptionView.setVisibility(VISIBLE);
      }
    }
  }

  public void configureSymbols(boolean useDarkSymbols) {
    if (useDarkSymbols) {
      imageCollapse = context.getResources().getDrawable(R.drawable.ic_action_collapse_dark);
      imageExpand = context.getResources().getDrawable(R.drawable.ic_action_expand_dark);

      Drawable imageForecastDark = context.getResources().getDrawable(R.drawable.ic_action_forecast_dark);
      forecastButton.setImageDrawable(imageForecastDark);
      Drawable imageDiagramDark = context.getResources().getDrawable(R.drawable.ic_action_picture_dark);
      diagramButton.setImageDrawable(imageDiagramDark);
    }
    setExpanded(expanded);
  }

  public void setTextColor(int color) {
    captionView.setTextColor(color);
  }

  public boolean isExpanded() {
    return expanded;
  }

  public void setExpanded(boolean expanded) {
    this.expanded = expanded;
    resizeButton.setImageDrawable(expanded ? imageCollapse : imageExpand);
    moreData.setVisibility(expanded ? View.VISIBLE : View.GONE);
  }
}

File: app/src/main/java/org/voegtle/weatherwidget/location/LocationView.kt
package org.voegtle.weatherwidget.location

import android.content.Context
import android.content.res.TypedArray
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.GridLayout
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import org.voegtle.weatherwidget.R
import org.voegtle.weatherwidget.data.Statistics
import org.voegtle.weatherwidget.data.StatisticsSet
import org.voegtle.weatherwidget.data.WeatherData
import org.voegtle.weatherwidget.databinding.ViewLocationBinding
import org.voegtle.weatherwidget.util.ColorUtil
import org.voegtle.weatherwidget.util.DataFormatter

class LocationView(private val currentContext: Context, attrs: AttributeSet) : LinearLayout(currentContext, attrs) {
    private var imageExpand = ContextCompat.getDrawable(currentContext, R.drawable.ic_action_expand)
    private var imageCollapse = ContextCompat.getDrawable(currentContext, R.drawable.ic_action_collapse)
    private var externalClickListener: OnClickListener? = null
    var diagramListener: OnClickListener? = null
    var forecastListener: OnClickListener? = null

    private lateinit var binding: ViewLocationBinding


    init {
        val li = currentContext.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        binding = ViewLocationBinding.inflate(li, this, true)

        val attributes = currentContext.theme.obtainStyledAttributes(attrs, R.styleable.LocationView, 0, 0)

        try {
            initializeTextViews(attributes)
            initializeButtons(attributes)
        } finally {
            attributes.recycle()
        }
    }

    private val formatter: DataFormatter = DataFormatter()

    var isExpanded: Boolean = false
        set(expanded) {
            field = expanded
            repaintExpandButton()
        }

    private fun repaintExpandButton() {
        binding.resizeButton.setImageDrawable(if (isExpanded) imageCollapse else imageExpand)
        binding.composeViewStatistics.visibility = if (isExpanded) View.VISIBLE else View.GONE
    }

    private fun initializeButtons(attributes: TypedArray) {
        binding.resizeButton.setOnClickListener { arg0 ->
            isExpanded = !isExpanded
            externalClickListener?.onClick(arg0)
        }
        repaintExpandButton()

        val diagrams = attributes.getBoolean(R.styleable.LocationView_diagrams, false)
        binding.diagramButton.visibility = if (diagrams) View.VISIBLE else View.GONE
        binding.diagramButton.setOnClickListener {
            diagramListener?.onClick(this@LocationView)
        }

        val forecast = attributes.getBoolean(R.styleable.LocationView_forecast, true)
        binding.forecastButton.visibility = if (forecast) View.VISIBLE else View.GONE
        binding.forecastButton.setOnClickListener { forecastListener?.onClick(this@LocationView) }
    }

    override fun setOnClickListener(l: OnClickListener?) {
        externalClickListener = l
    }

    private fun initializeTextViews(attributes: TypedArray) {
        binding.caption.text = attributes.getString(R.styleable.LocationView_caption)
    }

    fun setTextSize(textSize: Int) {
        binding.caption.textSize = (textSize + 1).toFloat()
        updateTextSize(binding.gridCurrent, textSize)
    }

    private fun updateTextSize(container: GridLayout, textSize: Int) {
        (0..container.childCount - 1)
            .map { container.getChildAt(it) as TextView }
            .forEach { it.textSize = textSize.toFloat() }
    }


    fun setCaption(captionText: String) {
        binding.caption.text = captionText
    }

    fun setData(data: WeatherData) {
        binding.temperature.text = formatter.formatTemperatureForActivity(data)
        binding.humidity.text = formatter.formatHumidityForActivity(data)

        setBarometer(data.barometer)
        setSolarradiation(data.solarradiation)
        setPowerProduction(data.powerProduction)
        setPowerFeed(data.powerFeed)
        setUVIndex(data.UV)

        setRainData(data.rain, binding.labelRainLastHour, binding.rainLastHour)
        setRainData(data.rainToday, binding.labelRainToday, binding.rainToday)
        setWind(data.wind)
        setWindGust(data.windgust);
        setSolarData(data.watt)
    }

    private fun setBarometer(value: Float?) {
        if (value != null && value > 0.0) {
            show(binding.labelBarometer, binding.barometer)
            binding.barometer.text = formatter.formatBarometer(value)
        } else {
            hide(binding.labelBarometer, binding.barometer)
        }
    }

    private fun setSolarradiation(value: Float?) {
        if (value != null && value > 0.0) {
            show(binding.labelSolarradiation, binding.solarradiation)
            binding.solarradiation.text = formatter.formatSolarradiation(value)
        } else {
            hide(binding.labelSolarradiation, binding.solarradiation)
        }
    }

    private fun setPowerProduction(powerProduction: Float?) {
        if (powerProduction != null && powerProduction >= 1.0) {
            show(binding.labelPowerProduction, binding.powerProduction)
            binding.powerProduction.text = formatter.formatWatt(powerProduction)
        } else {
            hide(binding.labelPowerProduction, binding.powerProduction)
        }
    }

    private fun setPowerFeed(powerFeed: Float?) {
        if (powerFeed != null && powerFeed >= 5.0) {
            show(binding.labelPowerFeed, binding.powerFeed)
            binding.powerFeed.text = formatter.formatWatt(powerFeed)
        } else {
            hide(binding.labelPowerFeed, binding.powerFeed)
        }
    }

    private fun setUVIndex(value: Float?) {
        if (value != null && value > 0.0) {
            show(binding.labelUv, binding.uv)
            binding.uv.text = formatter.formatInteger(value)
        } else {
            hide(binding.labelUv, binding.uv)
        }
    }

    private fun setSolarData(watt: Float?) {
        if (watt != null && watt > 0.0) {
            show(binding.labelSolarOutput, binding.solarOutput)
            binding.solarOutput.text = formatter.formatWatt(watt)
        } else {
            hide(binding.labelSolarOutput, binding.solarOutput)
        }
    }

    private fun setRainData(value: Float?, rainLabel: TextView, rain: TextView) {
        rainLabel.visibility = if (value != null && value > 0.0f) View.VISIBLE else View.GONE
        rain.visibility = if (value != null && value > 0.0f) View.VISIBLE else View.GONE
        rain.text = if (value != null && value > 0.0f) formatter.formatRain(value) else ""
    }

    private fun setWind(windgust: Float?) {
        if (windgust != null && windgust >= 1.0) {
            show(binding.labelWindSpeed, binding.windSpeed)
            binding.windSpeed.text = formatter.formatWind(windgust)
        } else {
            hide(binding.labelWindSpeed, binding.windSpeed)
        }
    }

    private fun setWindGust(windgust: Float?) {
        if (windgust != null && windgust >= 10.0) {
            show(binding.labelWindGust, binding.windGust)
            binding.windGust.text = formatter.formatWind(windgust)
        } else {
            hide(binding.labelWindGust, binding.windGust)
        }
    }

    fun setMoreData(statistics: Statistics) {
        binding.composeViewStatistics.setContent {
            StatisticsView(statistics)
        }
    }



    fun configureSymbols(useDarkSymbols: Boolean) {
        if (useDarkSymbols) {
            imageCollapse = ContextCompat.getDrawable(currentContext, R.drawable.ic_action_collapse_dark)
            imageExpand = ContextCompat.getDrawable(currentContext, R.drawable.ic_action_expand_dark)
            repaintExpandButton()

            val imageForecastDark = ContextCompat.getDrawable(currentContext, R.drawable.ic_action_forecast_dark)
            binding.forecastButton.setImageDrawable(imageForecastDark)
            val imageDiagramDark = ContextCompat.getDrawable(currentContext, R.drawable.ic_action_picture_dark)
            binding.diagramButton.setImageDrawable(imageDiagramDark)
        }
    }

    fun setTextColor(color: Int) {
        binding.caption.setTextColor(color)
    }

    fun highlight(highlight: Boolean) {
        if (highlight) {
            setBackgroundColor(ColorUtil.highlight())
        }
    }

    private fun hide(vararg view: TextView) {
        for (v in view) {
            v.visibility = View.GONE
        }
    }

    private fun show(vararg view: TextView) {
        for (v in view) {
            v.visibility = View.VISIBLE
        }
    }

}
